File: src/data/codes/single-layer-perceptron.m
clc;  % single-layer perceptron
clear;
close all;

%% =========================
% SETTINGS
% ==========================
useSynthetic = true;
csvFile = 'slp_loan_approval.csv';

%% =========================
% LOAD OR CREATE DATA
% ==========================
if useSynthetic
    rng(23);
    n = 160;

    Income_Level = randi([0 2], n, 1);       % Low=0, Medium=1, High=2
    Credit_Score = randi([0 2], n, 1);       % Low=0, Medium=1, High=2
    Employment_Status = randi([0 1], n, 1);  % Unstable=0, Stable=1
    Loan_Amount = randi([0 1], n, 1);        % Low=0, High=1

    score = 1.5*Income_Level + 1.5*Credit_Score + 1.2*Employment_Status - 1.0*Loan_Amount;
    Loan_Status = double(score >= 2.5);

    data = table(Income_Level, Credit_Score, Employment_Status, Loan_Amount, Loan_Status);
else
    data = readtable(csvFile);
end

disp(head(data));

%% PREPROCESSING
data = rmmissing(data);
data = unique(data);

X = [data.Income_Level, data.Credit_Score, data.Employment_Status, data.Loan_Amount]';
Y = data.Loan_Status';

%% TRAIN-TEST SPLIT
cv = cvpartition(Y, 'HoldOut', 0.2);
X_train = X(:, training(cv));
Y_train = Y(:, training(cv));
X_test  = X(:, test(cv));
Y_test  = Y(:, test(cv));

%% MODEL BUILDING
net = perceptron;
net = train(net, X_train, Y_train);

%% PREDICTION
Y_pred = net(X_test);
Y_pred = round(Y_pred);

%% EVALUATION
confMat = confusionmat(Y_test', Y_pred');
accuracy = sum(diag(confMat)) / sum(confMat(:));
TP = confMat(2,2); FP = confMat(1,2); FN = confMat(2,1);
precision = TP/(TP+FP+eps);
recall = TP/(TP+FN+eps);
f1 = 2*precision*recall/(precision+recall+eps);

disp('Confusion Matrix:');
disp(confMat);
fprintf('\nAccuracy  = %.4f\n', accuracy);
fprintf('Precision = %.4f\n', precision);
fprintf('Recall    = %.4f\n', recall);
fprintf('F1-score  = %.4f\n', f1);

figure;
confusionchart(Y_test', Y_pred');
title('Single Layer Perceptron Confusion Matrix');

File: src/data/codes/single-layer-perceptron.py
"""Single-layer perceptron (linear classifier)
# pip install numpy pandas scikit-learn matplotlib
"""
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import Perceptron
from sklearn.metrics import confusion_matrix, accuracy_score, precision_score, recall_score, f1_score
from sklearn.metrics import ConfusionMatrixDisplay
import matplotlib.pyplot as plt

USE_SYNTHETIC = True
RNG = np.random.default_rng(23)

if USE_SYNTHETIC:
    n = 160
    income = RNG.integers(0, 3, size=n)
    credit = RNG.integers(0, 3, size=n)
    employ = RNG.integers(0, 2, size=n)
    loan_amt = RNG.integers(0, 2, size=n)
    score = 1.5 * income + 1.5 * credit + 1.2 * employ - 1.0 * loan_amt
    status = (score >= 2.5).astype(int)
    data = pd.DataFrame(
        {
            "Income_Level": income,
            "Credit_Score": credit,
            "Employment_Status": employ,
            "Loan_Amount": loan_amt,
            "Loan_Status": status,
        }
    )
else:
    data = pd.read_csv("slp_loan_approval.csv")

print(data.head())
data = data.dropna().drop_duplicates()
X = data[["Income_Level", "Credit_Score", "Employment_Status", "Loan_Amount"]].values
y = data["Loan_Status"].values

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, stratify=y, random_state=42
)

model = Perceptron(max_iter=1000, random_state=42).fit(X_train, y_train)
y_pred = model.predict(X_test)

print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred))
print(f"Accuracy  = {accuracy_score(y_test, y_pred):.4f}")
print(f"Precision = {precision_score(y_test, y_pred, zero_division=0):.4f}")
print(f"Recall    = {recall_score(y_test, y_pred, zero_division=0):.4f}")
print(f"F1-score  = {f1_score(y_test, y_pred, zero_division=0):.4f}")

fig, ax = plt.subplots()
ConfusionMatrixDisplay.from_predictions(y_test, y_pred, ax=ax)
ax.set_title("Single-layer Perceptron Confusion Matrix")
plt.show()
